napari/layers/utils/plane.py

```diff
@@ -0,0 +1,166 @@
+from typing import Tuple
+
+import numpy as np
+from pydantic import validator
+
+from ...utils.events import EventedModel, SelectableEventedList
+from ...utils.geometry import intersect_line_with_plane_3d
+from ...utils.translations import trans
+
+
+class Plane(EventedModel):
+    """Defines a plane in 3D (with optional thickness) and provides utility methods
+    and events to handle it.
+
+    A plane is defined by a position, a normal vector and a thickness value, and can
+    be enabled or disabled.
+
+    Attributes
+    ----------
+    position : 3-tuple
+        A 3D position on the plane, defined in data coordinates.
+    normal : 3-tuple
+        A 3D unit vector normal to the plane, defined in data coordinates.
+    thickness : float
+        Thickness of the slice.
+    enabled : bool
+        Whether the plane is considered enabled.
+    """
+
+    position: Tuple[float, float, float] = (0, 0, 0)
+    normal: Tuple[float, float, float] = (1, 0, 0)
+    enabled: bool = False
```

Comment:
Should we enable by default? I feel like if you're making it, you probably want it enabled

Reply:
With clipping planes, true, and I agree in general this class should instantiate enabled planes. However, with @alisterburt 's `slicing_plane` we want it disabled by default (since it exists on the layer from instantiation). Easy to solve here (just add enabled=True there), but this might change depending on how we move forward with #3148 so it's worth noting.

Reply:
I think there's definitely an argument for having a plane class which we subclass as necessary for the different types (`SlicingPlane`, `ClippingPlane`) - thoughts? I know subclassing can be messy but this seems like the best way to go to me

Reply:
That's what I did before, and [people seemed to agree that it overcomplicates things](https://github.com/napari/napari/pull/3140#discussion_r688972201). Happy to go either way, but now I kinda like the fact that `Plane` is agnostic regarding its application.

Reply:
> subclass as necessary for the different types 

Yeah, I guess the question would be what's different for those types.  If it's just methods (not attributes), then like @brisvag says, there's an argument for keeping the model free of the business logic... and letting the thing that has the plane use it to clip, or whatever.  

But, I guess it depends on how they're different

Reply:
You're right that it's the same, removing the `Manager` and using the names `ClippingPlane`/`SlicingPlane` makes it feel a bit less 'imposing' for some reason - I'm happy whichever way this goes but do feel like these objects are different enough to warrant their own objects. 

I would be a little confused if I found a `Plane` object in the `ClippingPlaneList` which had a thickness and nothing happened when I changed it.

Another argument for separating them out is if we keep the reusable business logic in utility functions (which I support) we would have to differentiate between functions which operate on different types of the same object

For example
`intersect_line_with_plane()` (no thickness) and `intersect_line_with_slicing_plane` (with thickness), choosing the correct function as a user/dev requires knowledge about the other type of `Plane`, feels less obvious?

Reply:
I think we should keep in mind future plans:
1) ideas in #3148 do not play well with different objects
2) thickness will be useful for clipping as well if/when we implement soft clipping.

Reply:
> do feel like these objects are different enough to warrant their own objects.

well you've been using these objects much more than me, so I trust your intuition here.

I think this touches on a broader interesting topic about architecture.  I _think_ (though I'm not sure) that it would be nice if we had "largely declarative" models that were acted upon by rendering engines.  Layers are in many cases just a set of attributes, and it's up to the vispy classes (view layers) to turn those attributes into meaning. 

> we would have to differentiate between functions which operate on different types of the same object

This is a good argument.  I think there's a couple options here.  One is that you don't have both `intersect_line_with_plane()` (no thickness) and `intersect_line_with_slicing_plane` (with thickness) ... you just have `intersect_line_with_plane` and it checks if the plane has thickness.  

But, if you have a whole family of functions that _require_ thickness (and have no meaning without it), or a whole set of functions that expect the output of `intersect_line` to be another line, rather than a point... then I definitely agree with you that those are fundamentally different models that require different classes.   
(Indeed, a true plane and a "slab of space" are fundamentally different types).  And I'd soften/retract my comment from https://github.com/napari/napari/pull/3140#discussion_r688973818 

@sofroniewn?

Reply:
I'd be fine with `ClippingPlane` and `SlicingPlane` as two class names, and then `ClippingPlaneList`. Feels pretty natural and intuitive in a way that `PlaneManager` and `ThickPlaneManager` didn't before as @alisterburt noted above. 

If you'd like to move that over tomorrow before merge that's fine. @alisterburt and @brisvag you can decided this between yourselves during your timezone and feel free to have this PR merged by the time I get back online up in ~10 hours :-)

Reply:
Thanks for the input @tlambert03 @sofroniewn! I definitely like the idea of declarative models in general but dislike the 'forced' code separation that results: having functions which act on models rather than methods on objects forces users to think 'oh, where do I import that thing from again?'

I think this is my main argument against fully declarative models in a system that is designed to be easy for people to interact with programmatically... all of that to say 'I'm not sure which way I prefer overall' 🙂 

There are definitely a family of functions which require thickness, any 3d spatial stuff`point_in_slab_of_space()`/`volume_of_slab_of_space()` jumps to mind but I'm sure there are others

I'm going to PR this with the `ClippingPlane`/`SlicingPlane` changes soon, will merge once @brisvag has had a look at that

Reply:
>  'oh, where do I import that thing from again?'
>  I think this is my main argument against fully declarative models in a system that is designed to be easy for people to interact with programmatically..

oh definitely... one shouldn't have to import every utility function that they need.
but I don't think the two concepts are necessarily that mutually exclusive.  It's true that someone couldn't necessary use a method on a deeply nested object:  `layer.something.something.some_method()`, but that's where our main API design should come in and let someone use `layer.something()` ... which then dispatches to the appropriate utility functions (so that people don't have to remember where to import them from)

_or_ we start thinking outside pattern of having layers do everything, and start introducing new objects that manage these relationships between layers (i.e., a couple clipping planes and a couple layers)... and _that_ manager has an API that does all this.  (instead of each layer remembering the clipping planes that it relates to, and directly calling the methods on those plane objects).

>I'm going to PR this with the ClippingPlane/SlicingPlane changes soon, will merge once @brisvag has had a look at that

sounds good!  onward :)

Reply:
I still don't see the benefit of this split (`points_ins_slab_of_space` works just fine with thickness=0...), especially when considering the possible plans of using the same plane to (for example) clip one layer and slice another; with 2 different obects, we are forced to have both of them, and provide an api (and UI!) to "connect" them so they update together.

Anyways, I don't think this is important enough at the moment to hold up the merge, so we can go with the split for now. Just writing down my thoughts for when we revisit this!